File: src/Lessons/3-TimeComplexity/3.3-TapeEquilibrium.js
function solution(A){
    let array = [...A]; //to get code pred
    let totalSum = 0;
    let totalDeductionSum = 0;
    array.forEach(num => totalSum+=num);


    let minDiff;
    for (let index = 0; index < array.length-1; index++) {
        totalDeductionSum += array[index];
        totalSum -= array[index]
        console.log(totalDeductionSum+' :: '+totalSum);
        const tempDiff = Math.abs(totalDeductionSum - totalSum);

        if(minDiff == undefined) minDiff = tempDiff;

        if(tempDiff == 0) {
            minDiff = tempDiff;
            break;
        }
        if(tempDiff < minDiff) minDiff = tempDiff;
    }
    return minDiff;
}
const testArray = [3,1,2,4,3];
console.log(solution(testArray))
//expect 1

File: src/Lessons/3-TimeComplexity/3.3-TapeEquilibrium.go
package main

import (
	"fmt"
	"math"
)

func main() {

	fmt.Println(solution([]int{3, 1, 2, 4, 3}))
}

func solution(someArray []int) int {
	rightHandSum := 0 //store our sums
	leftHandSum := 0
	minDifference := -1
	//Get the total left hand sum
	for _, num := range someArray {
		rightHandSum += num
	}
	//start our loop from the last
	for index := 0; index < len(someArray); index++ {
		leftHandSum += someArray[index]
		rightHandSum -= someArray[index]
		tempDifference := int(math.Abs(float64(leftHandSum) - float64(rightHandSum)))

		if minDifference == -1 {
			minDifference = tempDifference
		}
		if minDifference == 0 {
			//Cant get lower than zero
			minDifference = tempDifference
			break
		}
		if minDifference > tempDifference {
			minDifference = tempDifference
		}
	}
	return minDifference
}
